src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs

```diff
@@ -81,7 +135,106 @@ public T DeserializePayload<T>(Message message)
     {
         var versionedMessage = message as VersionedMessage;
         var payloadSerializer = GetPayloadSerializer(versionedMessage?.Version);
-        return Deserialize<T>(payloadSerializer, message.Payload);
+
+        if (!DisableFastJson)
+        {
+            var messageWithRawMessage = message as MessageWithRawMessage;
+            var rawMessage = messageWithRawMessage.RawMessage;
+
+            if (payloadSerializer == s_payloadSerializer2)
+            {
+                // PERF: Fast path is compatibile only with protocol versions that use serializer_2,
+                // and this is faster than deserializing via deserializer_2.
+                var messageWithPayload = JsonConvert.DeserializeObject<PayloadedMessage<T>>(rawMessage, s_jsonSettings7);
+                return messageWithPayload.Payload;
+            }
+            else
+            {
+                // PERF: When payloadSerializer1 was resolved we need to deserialize JToken, and then deserialize that.
+                // This is still better than deserializing the JToken in DeserializeMessage because here we know that the payload
+                // will actually be used.
+                return Deserialize<T>(payloadSerializer, Deserialize<VersionedMessage>(rawMessage).Payload);
+            }
+        }
+        else
+        {
+            return Deserialize<T>(payloadSerializer, message.Payload);
+        }
+    }
+
+    private bool FastHeaderParse(string rawMessage, out int version, out string messageType)
+    {
+        // PERF: This can be also done slightly better using ReadOnlySpan<char> but we don't have that available by default in .NET Framework
+        // and the speed improvement does not warrant additional dependency. This is already taking just few ms for 10k messages.
+        version = 0;
+        messageType = null;
+
+        try
+        {
+            // {"Version":6,"MessageType":"TestExecution.GetTestRunnerProcessStartInfoForRunAll","Payload":{
+            if (rawMessage[2] != 'V')
```

Comment:
out of range check?

Reply:
If you only check for size 3 you will need to do a similar check for `10` and `11` later on so I would check at least this size.

Reply:
I am checking for minimal valid message size.